File: app/src/test/java/com/itba/runningMate/running/fragments/metrics/RunningMetricsPresenterTest.java
package com.itba.runningMate.running.fragments.metrics;

import com.itba.runningMate.domain.Route;
import com.itba.runningMate.domain.Run;
import com.itba.runningMate.repository.run.RunRepository;
import com.itba.runningMate.repository.runningstate.RunningStateStorage;
import com.itba.runningMate.services.location.Tracker;
import com.itba.runningMate.utils.Constants;
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider;
import com.itba.runningMate.utils.run.RunMetrics;

import org.junit.Before;
import org.junit.Test;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

import static com.itba.runningMate.utils.Constants.DISTANCE_EPSILON;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class RunningMetricsPresenterTest {

    private RunningStateStorage stateStorage;
    private RunRepository runRepository;
    private SchedulerProvider schedulers;
    private RunningMetricsView view;
    private Tracker tracker;

    private RunningMetricsPresenter presenter;
    private RunningMetricsPresenter presenterSpy;

    private Route route;

    @Before
    public void setUp() throws Exception {
        stateStorage = mock(RunningStateStorage.class);
        runRepository = mock(RunRepository.class);
        schedulers = mock(SchedulerProvider.class);
        view = mock(RunningMetricsView.class);
        tracker = mock(Tracker.class);

        presenter = new RunningMetricsPresenter(stateStorage, runRepository, schedulers, view);
        presenterSpy = spy(presenter);

        route = new Route().addToRoute(Constants.DEFAULT_LATITUDE, Constants.DEFAULT_LONGITUDE);
    }

    @Test
    public void givenViewAttachedThenAttachTrackingService() {
        presenter.onViewAttached();

        verify(view).attachTrackingService();
    }

    @Test
    public void givenViewDetachedThenPersistStateStorage() {
        presenter.onViewDetached();

        verify(stateStorage).persistState();
    }

    @Test
    public void givenViewDetachedThenDetachTrackingService() {
        presenter.onViewDetached();

        verify(view).detachTrackingService();
    }

    @Test
    public void givenStopWatchUpdateThenShowStopWatchUpdate() {
        long elapsedTime = 100L;

        presenter.onStopWatchUpdate(elapsedTime);

        verify(view).updateStopwatch(elapsedTime);
    }

    @Test
    public void givenDistanceUpdateThenShowUpdateDistance() {
        float elapsedDistance = 100L;

        presenter.onDistanceUpdate(elapsedDistance);

        verify(view).updateDistance(elapsedDistance);
    }

    @Test
    public void givenDistanceUpdateThenShowUpdateCalories() {
        float elapsedDistance = 100L;

        presenter.onDistanceUpdate(elapsedDistance);

        verify(view).updateCalories(RunMetrics.calculateCalories(elapsedDistance));
    }

    @Test
    public void givenPaceUpdateThenShowPaceUpdate() {
        long pace = 100L;

        presenter.onPaceUpdate(pace);

        verify(view).updatePace(pace);
    }

    @Test
    public void givenStopButtonClickWhenDistanceLessThanThresholdTHenShowStopConfirm() {
        when(tracker.queryDistance()).thenReturn((float) (DISTANCE_EPSILON - 0.5));

        presenter.onTrackingServiceAttached(tracker);
        presenter.onStopButtonClick();

        verify(view).showStopConfirm();
    }

    @Test
    public void givenPlayButtonClickThenResumeTracking() {
        when(tracker.isTracking()).thenReturn(false);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPlayButtonClick();

        verify(tracker).newLap();
        verify(tracker).resumeTracking();
    }

    @Test
    public void givenPlayButtonClickThenHidePlayButton() {
        when(tracker.isTracking()).thenReturn(false);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPlayButtonClick();

        verify(view).hidePlayBtn();
    }

    @Test
    public void givenPlayButtonClickThenHideStopButton() {
        when(tracker.isTracking()).thenReturn(false);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPlayButtonClick();

        verify(view).hideStopBtn();
    }

    @Test
    public void givenPlayButtonClickThenShowPauseButton() {
        when(tracker.isTracking()).thenReturn(false);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPlayButtonClick();

        verify(view).showPauseBtn();
    }


    @Test
    public void givenPauseButtonClickClickThenStopTracking() {
        when(tracker.isTracking()).thenReturn(true);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPauseButtonClick();

        verify(tracker).stopTracking();
    }


    @Test
    public void givenPauseButtonClickClickThenHidePauseButton() {
        when(tracker.isTracking()).thenReturn(true);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPauseButtonClick();

        verify(view).hidePauseBtn();
    }

    @Test
    public void givenPauseButtonClickClickThenShowPlayButton() {
        when(tracker.isTracking()).thenReturn(true);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPauseButtonClick();

        verify(view).showPlayBtn();
    }

    @Test
    public void givenPauseButtonClickClickThenHideStopButton() {
        when(tracker.isTracking()).thenReturn(true);

        presenter.onTrackingServiceAttached(tracker);
        presenter.onPauseButtonClick();

        verify(view).showStopBtn();
    }

    @Test
    public void givenTrackingServiceDetachedThenSetIsTrackerAttachedFalse() {
        presenter.onTrackingServiceDetached();

        assertFalse(presenter.isTrackerAttached());
    }

    @Test
    public void givenTrackingServiceDetachedThenSetTrackerNull() {
        presenter.onTrackingServiceDetached();

        assertNull(presenter.getTracker());
    }

    @Test
    public void givenTrackingServiceAttachedThenSetTrackingLocationUpdateListener() {
        presenter.onTrackingServiceAttached(tracker);

        verify(tracker).setTrackingMetricsUpdateListener(presenter);
    }

    @Test
    public void givenTrackingServiceAttachedThenShowTrackedMetrics() {
        long pace = 10L;
        float distance = 40F;
        long elapsedTime = 1000L;
        when(tracker.isTracking()).thenReturn(true);
        when(tracker.queryDistance()).thenReturn(distance);
        when(tracker.queryPace()).thenReturn(pace);
        when(tracker.queryElapsedTime()).thenReturn(elapsedTime);

        presenterSpy.onTrackingServiceAttached(tracker);


        verify(presenterSpy).onPaceUpdate(pace);
        verify(presenterSpy).onDistanceUpdate(distance);
        verify(presenterSpy).onStopWatchUpdate(elapsedTime);
    }

    @Test
    public void givenStopRunWhenTrackingThenStopTracking() {
        presenter.onTrackingServiceAttached(tracker);

        presenter.stopRun();

        verify(tracker).stopTracking();
    }

    @Test
    public void givenStopRunWhenTrackingDistanceLessThanThresholdThenEndActivity() {
        when(tracker.queryDistance()).thenReturn((float) (DISTANCE_EPSILON - 0.5));

        presenter.onTrackingServiceAttached(tracker);
        presenter.stopRun();

        verify(view).finishActivity();
    }

    @Test
    public void givenStopRunWhenTrackingDistanceGreaterThanThresholdThenSaveRun() {
        when(tracker.queryDistance()).thenReturn(10F);
        when(tracker.queryElapsedTime()).thenReturn(10000L);
        when(tracker.queryStartTime()).thenReturn(System.currentTimeMillis());
        when(tracker.queryEndTime()).thenReturn(System.currentTimeMillis() + 10000L);
        when(tracker.queryVelocity()).thenReturn(10F / 10000F);
        when(tracker.queryPace()).thenReturn(10000L / 10);
        when(tracker.queryRoute()).thenReturn(route);

        when(schedulers.io()).thenReturn(Schedulers.trampoline());
        when(schedulers.ui()).thenReturn(Schedulers.trampoline());

        long runId = 1L;
        when(runRepository.insertRun(any(Run.class))).thenReturn(Single.just(runId));

        presenter.onTrackingServiceAttached(tracker);
        presenter.stopRun();

        verify(view).launchRunActivity(runId);
    }

    @Test
    public void givenStopRunWhenTrackingDistanceGreaterThanThresholdAndErrorOnSaveRunThenShowErrorMessage() {
        when(tracker.queryDistance()).thenReturn(10F);
        when(tracker.queryElapsedTime()).thenReturn(10000L);
        when(tracker.queryStartTime()).thenReturn(System.currentTimeMillis());
        when(tracker.queryEndTime()).thenReturn(System.currentTimeMillis() + 10000L);
        when(tracker.queryVelocity()).thenReturn(10F / 10000F);
        when(tracker.queryPace()).thenReturn(10000L / 10);
        when(tracker.queryRoute()).thenReturn(route);

        when(schedulers.io()).thenReturn(Schedulers.trampoline());
        when(schedulers.ui()).thenReturn(Schedulers.trampoline());

        long runId = 1L;
        when(runRepository.insertRun(any(Run.class))).thenReturn(Single.error(new RuntimeException("Could not save run")));

        presenter.onTrackingServiceAttached(tracker);
        presenter.stopRun();

        verify(view).showSaveRunError();
    }

}

File: app/src/test/java/com/itba/runningMate/running/fragments/metrics/RunningMetricsPresenterTest.kt
package com.itba.runningMate.running.fragments.metrics

import com.google.firebase.messaging.FirebaseMessaging
import com.itba.runningMate.domain.AggregateRunMetricsDetail
import com.itba.runningMate.domain.Route
import com.itba.runningMate.domain.Run
import com.itba.runningMate.repository.achievements.AchievementsRepository
import com.itba.runningMate.repository.aggregaterunmetrics.AggregateRunMetricsStorage
import com.itba.runningMate.repository.run.RunRepository
import com.itba.runningMate.services.location.Tracker
import com.itba.runningMate.utils.Constants
import com.itba.runningMate.utils.Constants.DISTANCE_EPSILON
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider
import com.itba.runningMate.utils.run.RunMetrics.calculateCalories
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.argumentCaptor

class RunningMetricsPresenterTest {

    private lateinit var firebaseMessaging: FirebaseMessaging
    private lateinit var aggregateRunMetricsStorage: AggregateRunMetricsStorage
    private lateinit var achievementsRepository: AchievementsRepository
    private lateinit var runRepository: RunRepository
    private lateinit var schedulers: SchedulerProvider
    private lateinit var view: RunningMetricsView
    private lateinit var tracker: Tracker
    private lateinit var presenter: RunningMetricsPresenter
    private lateinit var presenterSpy: RunningMetricsPresenter
    private lateinit var route: Route

    private var aggregateMetrics: AggregateRunMetricsDetail = AggregateRunMetricsDetail.Builder()
        .distance(50f)
        .calories(40)
        .pace(40)
        .runningTime(40)
        .elapsedTime(40)
        .speed(30f)
        .build()

    @Before
    @Throws(Exception::class)
    fun setUp() {
        firebaseMessaging = Mockito.mock(FirebaseMessaging::class.java)
        aggregateRunMetricsStorage = Mockito.mock(AggregateRunMetricsStorage::class.java)
        achievementsRepository = Mockito.mock(AchievementsRepository::class.java)
        runRepository = Mockito.mock(RunRepository::class.java)
        schedulers = Mockito.mock(SchedulerProvider::class.java)
        view = Mockito.mock(RunningMetricsView::class.java)
        tracker = Mockito.mock(Tracker::class.java)
        presenter = RunningMetricsPresenter(
            runRepository,
            schedulers,
            achievementsRepository,
            aggregateRunMetricsStorage,
            firebaseMessaging,
            view
        )
        presenterSpy = Mockito.spy(presenter)
        route = Route().addToRoute(Constants.DEFAULT_LATITUDE, Constants.DEFAULT_LONGITUDE)
    }

    @Test
    fun givenViewAttachedThenAttachTrackingService() {
        presenter.onViewAttached()
        Mockito.verify(view).attachTrackingService()
    }

    @Test
    fun givenViewDetachedThenDetachTrackingService() {
        presenter.onViewDetached()
        Mockito.verify(view).detachTrackingService()
    }

    @Test
    fun givenStopWatchUpdateThenShowStopWatchUpdate() {
        val elapsedTime = 100L
        presenter.onStopWatchUpdate(elapsedTime)
        Mockito.verify(view).updateStopwatch(elapsedTime)
    }

    @Test
    fun givenDistanceUpdateThenShowUpdateDistance() {
        val elapsedDistance = 100f
        presenter.onDistanceUpdate(elapsedDistance)
        Mockito.verify(view).updateDistance(elapsedDistance)
    }

    @Test
    fun givenDistanceUpdateThenShowUpdateCalories() {
        val elapsedDistance = 100f
        presenter.onDistanceUpdate(elapsedDistance)
        Mockito.verify(view).updateCalories(calculateCalories(elapsedDistance))
    }

    @Test
    fun givenPaceUpdateThenShowPaceUpdate() {
        val pace = 100L
        presenter.onPaceUpdate(pace)
        Mockito.verify(view).updatePace(pace)
    }

    @Test
    fun givenStopButtonClickWhenDistanceLessThanThresholdTHenShowStopConfirm() {
        Mockito.`when`(tracker.queryDistance()).thenReturn((DISTANCE_EPSILON - 0.5).toFloat())
        presenter.onTrackingServiceAttached(tracker)
        presenter.onStopButtonClick()
        Mockito.verify(view).showStopConfirm()
    }

    @Test
    fun givenPlayButtonClickThenResumeTracking() {
        Mockito.`when`(tracker.isTracking()).thenReturn(false)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPlayButtonClick()
        Mockito.verify(tracker).newLap()
        Mockito.verify(tracker).resumeTracking()
    }

    @Test
    fun givenPlayButtonClickThenHidePlayButton() {
        Mockito.`when`(tracker.isTracking()).thenReturn(false)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPlayButtonClick()
        Mockito.verify(view).hidePlayBtn()
    }

    @Test
    fun givenPlayButtonClickThenHideStopButton() {
        Mockito.`when`(tracker.isTracking()).thenReturn(false)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPlayButtonClick()
        Mockito.verify(view).hideStopBtn()
    }

    @Test
    fun givenPlayButtonClickThenShowPauseButton() {
        Mockito.`when`(tracker.isTracking()).thenReturn(false)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPlayButtonClick()
        Mockito.verify(view).showPauseBtn()
    }

    @Test
    fun givenPauseButtonClickClickThenStopTracking() {
        Mockito.`when`(tracker.isTracking()).thenReturn(true)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPauseButtonClick()
        Mockito.verify(tracker).stopTracking()
    }

    @Test
    fun givenPauseButtonClickClickThenHidePauseButton() {
        Mockito.`when`(tracker.isTracking()).thenReturn(true)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPauseButtonClick()
        Mockito.verify(view).hidePauseBtn()
    }

    @Test
    fun givenPauseButtonClickClickThenShowPlayButton() {
        Mockito.`when`(tracker.isTracking()).thenReturn(true)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPauseButtonClick()
        Mockito.verify(view).showPlayBtn()
    }

    @Test
    fun givenPauseButtonClickClickThenHideStopButton() {
        Mockito.`when`(tracker.isTracking()).thenReturn(true)
        presenter.onTrackingServiceAttached(tracker)
        presenter.onPauseButtonClick()
        Mockito.verify(view).showStopBtn()
    }

    @Test
    fun givenTrackingServiceDetachedThenSetIsTrackerAttachedFalse() {
        presenter.onTrackingServiceDetached()
        Assert.assertFalse(presenter.isTrackerAttached)
    }

    @Test
    fun givenTrackingServiceDetachedThenSetTrackerNull() {
        presenter.onTrackingServiceDetached()
        Assert.assertNull(presenter.tracker)
    }

    @Test
    fun givenTrackingServiceAttachedThenSetTrackingLocationUpdateListener() {
        presenter.onTrackingServiceAttached(tracker)
        Mockito.verify(tracker).setTrackingMetricsUpdateListener(presenter)
    }

    @Test
    fun givenTrackingServiceAttachedThenShowTrackedMetrics() {
        val pace = 10L
        val distance = 40f
        val elapsedTime = 1000L
        Mockito.`when`(tracker.isTracking()).thenReturn(true)
        Mockito.`when`(tracker.queryDistance()).thenReturn(distance)
        Mockito.`when`(tracker.queryPace()).thenReturn(pace)
        Mockito.`when`(tracker.queryElapsedTime()).thenReturn(elapsedTime)
        presenterSpy.onTrackingServiceAttached(tracker)
        Mockito.verify(presenterSpy).onPaceUpdate(pace)
        Mockito.verify(presenterSpy).onDistanceUpdate(distance)
        Mockito.verify(presenterSpy).onStopWatchUpdate(elapsedTime)
    }

    @Test
    fun givenStopRunWhenTrackingThenStopTracking() {
        presenter.onTrackingServiceAttached(tracker)
        presenter.stopRun()
        Mockito.verify(tracker).stopTracking()
    }

    @Test
    fun givenStopRunWhenTrackingDistanceLessThanThresholdThenEndActivity() {
        Mockito.`when`(tracker.queryDistance()).thenReturn((DISTANCE_EPSILON - 0.5).toFloat())
        presenter.onTrackingServiceAttached(tracker)
        presenter.stopRun()
        Mockito.verify(view).finishActivity()
    }

    @Test
    fun givenStopRunWhenTrackingDistanceGreaterThanThresholdThenSaveRun() {
        Mockito.`when`(tracker.queryDistance()).thenReturn(10f)
        Mockito.`when`(tracker.queryElapsedTime()).thenReturn(10000L)
        Mockito.`when`(tracker.queryStartTime()).thenReturn(System.currentTimeMillis())
        Mockito.`when`(tracker.queryEndTime()).thenReturn(System.currentTimeMillis() + 10000L)
        Mockito.`when`(tracker.queryVelocity()).thenReturn(10f / 10000f)
        Mockito.`when`(tracker.queryPace()).thenReturn(10000L / 10)
        Mockito.`when`(tracker.queryRoute()).thenReturn(route)
        Mockito.`when`(aggregateRunMetricsStorage.getAggregateRunMetricsDetail())
            .thenReturn(aggregateMetrics)

        Mockito.`when`(schedulers.io()).thenReturn(Schedulers.trampoline())
        Mockito.`when`(schedulers.ui()).thenReturn(Schedulers.trampoline())
        Mockito.`when`(schedulers.computation()).thenReturn(Schedulers.trampoline())

        val runId = 1L
        val argumentCaptor = argumentCaptor<Run>()

        Mockito.`when`(
            runRepository.insertRun(argumentCaptor.capture())
        ).thenReturn(Single.just(runId))

        Mockito.`when`(
            achievementsRepository.addAchievements(any(), any())
        ).thenReturn(Completable.complete())

        presenter.onTrackingServiceAttached(tracker)
        presenter.stopRun()
        Mockito.verify(view).launchRunActivity(runId)
    }

    @Test
    fun givenStopRunWhenTrackingDistanceGreaterThanThresholdAndErrorOnSaveRunThenShowErrorMessage() {
        Mockito.`when`(tracker.queryDistance()).thenReturn(10f)
        Mockito.`when`(tracker.queryElapsedTime()).thenReturn(10000L)
        Mockito.`when`(tracker.queryStartTime()).thenReturn(System.currentTimeMillis())
        Mockito.`when`(tracker.queryEndTime()).thenReturn(System.currentTimeMillis() + 10000L)
        Mockito.`when`(tracker.queryVelocity()).thenReturn(10f / 10000f)
        Mockito.`when`(tracker.queryPace()).thenReturn(10000L / 10)
        Mockito.`when`(tracker.queryRoute()).thenReturn(route)
        Mockito.`when`(schedulers.io()).thenReturn(Schedulers.trampoline())
        Mockito.`when`(schedulers.ui()).thenReturn(Schedulers.trampoline())
        Mockito.`when`(schedulers.computation()).thenReturn(Schedulers.trampoline())

        Mockito.`when`(aggregateRunMetricsStorage.getAggregateRunMetricsDetail())
            .thenReturn(aggregateMetrics)

        Mockito.`when`(
            achievementsRepository.addAchievements(any(), any())
        ).thenReturn(Completable.complete())

        Mockito.`when`(
            runRepository.insertRun(any())
        ).thenReturn(Single.error(RuntimeException()))

        presenter.onTrackingServiceAttached(tracker)
        presenter.stopRun()

        Mockito.verify(view).showSaveRunError()
    }
}